fvm/src/syscalls/event.rs

```diff
@@ -0,0 +1,14 @@
+use fvm_shared::event::ActorEvent;
+
+use super::Context;
+use crate::kernel::Result;
+use crate::Kernel;
+
+pub fn emit_event(
+    context: Context<'_, impl Kernel>,
+    event_off: u32, // ActorEvent
+    event_len: u32,
+) -> Result<()> {
+    let evt: ActorEvent = context.memory.read_cbor(event_off, event_len)?;
+    context.kernel.emit_event(evt)
```

Comment:
So, um. I'm going to make you hate me. We should do this with a gas-charging serde visitor. But we can do that in a followup patch (I'm "happy" to write it).

Basically, we can charge gas _as_ we deserialize and stop immediately if/when we run out without ever doing any work we haven't charged for.

Reply:
(we can probably extend this concept to other `read_cbor` cases)

Reply:
We already have these gas fees in place here:

1. Syscall fee, which we had discussed to extend with dynamic prices based on parameter lengths.
2. Event emission -- flat rate per event.
3. Event emission -- per byte cost of serialized event. This is a storage cost.
4. Event emission -- per indexed field.

In a **_perfect_** world, we would charge for the cost of serde by metering it -- we have discussed about instrumenting codec logic and charging gas for it, but my impression is that that's a _tomorrow_ thing.

In absence of the above, what would the pricing policy of this gas-charging serde visitor be? Would we charge for anything other than bytes consumed? Can't we include the projected cost for that in (1) or (3)?

Reply:
The problem with the current version is that we charge _after_ processing the event. If we use a serde visitor, se'd be able to charge for 2-4 while parsing (and stop parsing early if we run out of gas).

Reply:
Would the syscall fee with parameter length costs, charged at the time of the syscall, cover the concern here?

What I'm trying to figure out is what exactly are we charging for with this gas-charging serde visitor, if we are not charging for compute? Aren't we just charging for bytes (which can be covered in a simpler manner through (1))?

Reply:
We would be charging for compute.

To be clear, I'm not suggesting that we change the charging model. We're _already_ charging for all the operations in 2-4. However, a user can get us to do a lot of free work before we get around to charging because we have to first _parse_ the structure entirely before we can determine how much it should cost.

I'm just trying to find a way to charge _as_ we parse, so we can abort early.

But this isn't even consensus critical, so we can punt till later.